network/simple-onchain-discovery/src/lib.rs

```diff
@@ -215,3 +234,107 @@ impl ConfigurationChangeListener {
         );
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use diem_config::config::HANDSHAKE_VERSION;
+    use diem_crypto::{
+        ed25519::{Ed25519PrivateKey, Ed25519PublicKey},
+        x25519::PrivateKey,
+        PrivateKey as PK, Uniform,
+    };
+    use diem_types::{
+        on_chain_config::OnChainConfig, validator_config::ValidatorConfig,
+        validator_info::ValidatorInfo, PeerId,
+    };
+    use futures::executor::block_on;
+    use rand::{rngs::StdRng, SeedableRng};
+    use tokio::{
+        runtime::Runtime,
+        time::{timeout_at, Duration},
+    };
+
+    #[test]
+    fn restart_if_key_mismatch() {
+        diem_logger::DiemLogger::init_for_testing();
+        let runtime = Runtime::new().unwrap();
+        let consensus_private_key = Ed25519PrivateKey::generate_for_testing();
+        let consensus_pubkey = consensus_private_key.public_key();
+        let pubkey = test_pubkey([0u8; 32]);
+        let different_pubkey = test_pubkey([1u8; 32]);
+        let peer_id = PeerId::from_identity_public_key(pubkey);
+
+        // Build up the Reconfig Listener
+        let (conn_mgr_reqs_tx, _rx) = channel::new_test(1);
+        let (mut reconfig_tx, reconfig_rx) = gen_simple_discovery_reconfig_subscription();
+        let listener = ConfigurationChangeListener::new(
+            NetworkContext::mock_with_peer_id(peer_id),
+            pubkey,
+            Encryptor::for_testing(),
+            conn_mgr_reqs_tx,
+            reconfig_rx,
+        );
+
+        // Build up and send an update with a different pubkey
+        send_pubkey_update(
+            peer_id,
+            consensus_pubkey,
+            different_pubkey,
+            &mut reconfig_tx,
+        );
+
+        let listener_future = async move {
+            // Run the test, ensuring we actually stop after a couple seconds in case it fails to fail
+            timeout_at(
+                tokio::time::Instant::from(Instant::now() + Duration::from_secs(2)),
+                listener.start(),
+            )
+            .await
+            .expect("Shouldn't timeout");
+        };
+
+        // Ensure we get the right panic
+        if let Err(err) = block_on(runtime.spawn(listener_future)) {
+            assert_eq!(
+                "Onchain pubkey differs from local pubkey",
+                *err.into_panic().downcast::<&'static str>().unwrap()
+            );
```

Comment:
I think you can decorate the test to expect a specific panic message like:

```rust
#[test]
#[should_panic(expected = "")]
fn restart_if_key_mismatch() {
    // ..
    block_on(listener_future);
}
```

then the test will fail if it doesn't panic or the message doesn't match : )

Reply:
Originally, I was doing a join which was wrapping the panic (why I did it this way), so I probably don't have to do it anymore, now that I don't have the join.